### PR TITLE
Adds a when() method to the query builder.

### PR DESCRIPTION
**Pull request by lonnieezell**
**Description**

This adds a method, `when()` to the Query Builder that can conditionally modify queries based on the truthiness of a given condition. 

**Checklist:**
- [X] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by kenjis** [comment]
This is an enhancement, so it should go to `4.3` branch, not `develop`.

**Comment by kenjis** [comment]
@lonnieezell Only changing the base branch on GitHub will not work.
Because you created the PR branch based on `develop`.

Even if you rebase, if there are commits that have not been merged into `4.3`, the PR branch will contain them.

The easy way is to create a new branch from `4.3` and cherry-pick commits that are needed.

**Comment by datamweb** [comment]
@kenjis , this is a question for me too, if the user has chosen the wrong branch and opens PR. How can he correct it? Please add the necessary explanations to file [workflow](https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md) if possible.

**Comment by kenjis** [comment]
@datamweb 
If you have the PR branch `feat-abc`:
1. Create a new branch `feat-abc-new` from the correct branch, and cherry-pick the commits you did.
2. Delete the PR branch `feat-abc`, and rename the new branch  `feat-abc-new` to `feat-abc`.
3. Force push.
4. On GitHub PR page, change the base branch to the correct branch.

**Comment by datamweb** [comment]
Great, got it. 
**Thanks**.

**Comment by kenjis** [comment]
@datamweb I sent a PR #6372

**Comment by lonnieezell** [comment]
@kenjis fixed to point toward 4.3 now

File: user_guide_src/source/changelogs/v4.2.2.rst [hunk 1]
```diff
@@ -16,7 +16,12 @@ BREAKING
 - The method signature of ``CodeIgniter\Debug\Exceptions::__construct()`` has been changed. The ``IncomingRequest`` typehint on the ``$request`` parameter was removed. Extending classes should likewise remove the parameter so as not to break LSP.
 - The method signature of ``BaseBuilder.php::insert()`` and ``BaseBuilder.php::update()`` have been changed. The ``?array`` typehint on the ``$set`` parameter was removed.
 - A bug that caused pages to be cached before after filters were executed when using page caching has been fixed. Adding response headers or changing the response body in after filters now caches them correctly.
-- Due to a bug fix, now :php:func:`random_string()` with the first parameter ``'crypto'`` throws ``InvalidArgumentException`` if the second parameter ``$len`` is an odd number.
+- Due to a bug fix, now :php:func:`random_string()` with the first parameter 'crypto' throws InvalidArgumentException if the second parameter $len is an odd number.
+
+Enhancements
+************
+
+none.
```
Comment (kenjis) [inline user_guide_src/source/changelogs/v4.2.2.rst:24]: Remove this change.

Comment (kenjis) [inline user_guide_src/source/changelogs/v4.2.2.rst:24]: Remove this change.

File: system/Database/BaseBuilder.php [hunk 2]
```diff
@@ -2422,6 +2422,22 @@ public function decrement(string $column, int $value = 1)
         return true;
     }
 
+    /**
+     * Only runs the query when $condition evaluates to true
+     *
+     * @param mixed $condition
```
Comment (kenjis) [inline system/Database/BaseBuilder.php:2428]: Don't use `mixed`. See #6310

File: tests/system/Database/Builder/WhenTest.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+/**
+ * @internal
+ */
+final class WhenTest extends CIUnitTestCase
+{
+    /**
+     * @var MockConnection
+     */
+    protected $db;
```
Comment (kenjis) [inline tests/system/Database/Builder/WhenTest.php:25]: Use property type, instead of `@var`.

**Comment by kenjis** [comment]
Please add changelog and link to the details page:
https://github.com/codeigniter4/CodeIgniter4/blob/4.3/user_guide_src/source/changelogs/v4.3.0.rst#enhancements

**Comment by iRedds** [comment]
What if this code is implemented as a trait?
It does not depend on the query builder, so it can also be used, for example, on the Response class or model.

File: tests/system/Database/Builder/WhenTest.php (hunk 3 above)
Comment (lonnieezell) [reply tests/system/Database/Builder/WhenTest.php:25]: @kenjis I'm not sure exactly what you're looking for here....

Comment (kenjis) [inline tests/system/Database/Builder/WhenTest.php:25]: I mean like this:
```suggestion
    protected MockConnection $db;
```

File: user_guide_src/source/changelogs/v4.2.2.rst (hunk 1 above)
Comment (lonnieezell) [reply user_guide_src/source/changelogs/v4.2.2.rst]: Done

File: system/Database/BaseBuilder.php (hunk 2 above)
Comment (lonnieezell) [reply system/Database/BaseBuilder.php]: Done

File: tests/system/Database/Builder/WhenTest.php (hunk 3 above)
Comment (lonnieezell) [reply tests/system/Database/Builder/WhenTest.php:25]: Oh, silly me. Thanks for clarification.

**Comment by lonnieezell** [comment]
Well, looks like pulling in the changes from remote screwed me over here.  Will fix that tomorrow.

**Comment by iRedds** [comment]
I was sure I left a comment on this PR, but for some reason I didn't find it. I must have gotten too old. =)

I would like to suggest reconsidering the idea of this feature.

The ``when`` method is not logically related to databases and can be applied to the framework API.
It seems to me that it would be more convenient to put this method into a separate trait, which can be added to such classes as Request, Response, Model, and other classes.

In addition to the ``when`` method, you can also add the ``whenNot`` method.

**Comment by lonnieezell** [comment]
> The when method is not logically related to databases and can be applied to the framework API.
It seems to me that it would be more convenient to put this method into a separate trait, which can be added to such classes as Request, Response, Model, and other classes.

In addition to the when method, you can also add the whenNot method.

@iRedds I think that's a great idea. Will do.

**Comment by lonnieezell** [comment]
Closing this as it's been replaced by #6574

**Pull request closed without merging**